<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" class="global_function_countkmers" data-page="countKmers">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=US-ASCII" />
    <title>countKmers - SeqAn API Documentation</title>

      

      
      <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

      <link rel="stylesheet" href="lib/bootstrap/css/bootstrap.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap/js/bootstrap.min.js"></script>

      <link rel="stylesheet" href="lib/bootstrap-multiselect/css/bootstrap-multiselect.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap-multiselect/js/bootstrap-multiselect.js"></script>

      <script type="text/javascript" charset="utf-8" src="lib/ZeroClipboard/ZeroClipboard.min.js"></script>

      <link rel="stylesheet" href="css/style.css" type="text/css" media="screen" charset="utf-8" />
      <link rel="stylesheet" href="css/common.less.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="js/less.min.js"></script>
      



    <script type="text/javascript" charset="utf-8">
      hasFrames = window.top.frames.main ? true : false;
      relpath = '/';
      docsPrefix = 'docs/yard';
      listPrefix = 'list/docs/yard';
      searchPrefix = 'search/docs/yard';
      framesUrl = '/docs/yard/frames/file/README.md';
    </script>

    <style type="text/css">
    <!-- pygments not available -->
    </style>

    <style>
      .link.error
      {
          color: red;
      }
    </style>

      
      
      <script type="text/javascript" charset="utf-8" src="lib/jquery-bbq/jquery.ba-bbq.min.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/jquery.smooth-scroll.js"></script>

      <script type="text/javascript" charset="utf-8" src="js/lang_entities.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/app.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/autocomplete.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/rubydoc_custom.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/common.js"></script>
      
  </head>
  <body>
    <a id="top" name="top"></a>

    <div id="content">
      
<h1 data-lang-entity="function" data-pimped="true">
  <a href="page_LanguageEntities.html#function">fn()</a> <span>countKmers<div><div>Counts kmers in a sequence. Optionally, a background model is returned.</div></div></span>
</h1>


<table class="overview">

  

  

  

  

  
  </tr>

  <tr>
    <th>Defined in</th>
    <td>
      
      &lt;seqan/alignment_free.h&gt;
      
      
    </td>
  </tr>
  
  <tr>
    <th>Signature</th>
    <td>
      <code>void countKmers(kmerCounts, sequence, k);
void countKmers(kmerCounts, bgFrequencies, sequence, k);
void countKmers(kmerCounts, bgModel, sequence, k);
</code>
    </td>
  </tr>
</table>


<!-- @param -->



<h2>Parameters</h2>

<table class="overview">

    <tr>
        <th>
        
            
                <code data-param-type="out" title="OUT parameters are modified and used as a means to return values">kmerCounts</code>
            
        
        </th>
        <td><a href="class_String.html" data-lang-entity="class">String</a> of <tt>unsigned</tt> with kmer counts for every k-mer.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="out" title="OUT parameters are modified and used as a means to return values">bgFrequencies</code>
            
        
        </th>
        <td><a href="class_String.html" data-lang-entity="class">String</a> of background frequencies (<tt>double</tt>) representing the model.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="out" title="OUT parameters are modified and used as a means to return values">bgModel</code>
            
        
        </th>
        <td><a href="class_MarkovModel.html" data-lang-entity="class">MarkovModel</a> to use.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">sequence</code>
            
        
        </th>
        <td><a href="class_String.html" data-lang-entity="class">String</a> (sequence) where k-mers are counted.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">k</code>
            
        
        </th>
        <td>k-mer length (<tt>unsigned</tt>).</td>
    </tr>

</table>




<!-- @tparam -->





<!-- @returns -->




<!--
 =============================================================================
  Detailed Description
 =============================================================================
-->


<h2>Detailed Description</h2>
<div class="docstring">
  <div class="discussion">
    <!-- @internal -->
    

    <!-- @deprecated -->
    

    <!-- @warning -->
    

    <!-- @note -->
    

    <!-- @aka -->
    

    <!-- paragraphs -->
    
<div><p>k-mers overlapping masked (aka 'N') letters are not counted in case of Dna5Strings. A Bernoulli or Markov Model
can be choosen as a background model.</p><h3>Examples</h3><p>Calculate the alignment free sequence similarity o two masked DNA sequences.</p><div data-src-path="None"><pre class="code">using namespace seqan;
// Masked sequence, we do not want to count words overlapping 'N'
Dna5String sequenceDna5 =
    "TAGGTTTTCCGAAAAGGTAGCAACTTTACGTGATCAAACCTCTGACGGGGTTTTCCCCGTCGAAATTGGGTG"
    "TTTCTTGTCTTGTTCTCACTTGGGGCATCTCCGTCAAGCCAAGAAAGTGCTCCCTGGATTCTGTTGCTAACG"
    "AGTCTCCTCTGCATTCCTGCTTGACTGATTGGGCGGACGGGGTGTCCACCTGACGCTGAGTATCGCCGTCAC"
    "GGTGCCACATGTCTTATCTATTCAGGGATCAGAATTTATTCAGGAAATCAGGAGATGCTACACTTGGGTTAT"
    "CGAAGCTCCTTCCAAGGCGTAGCAAGGGCGACTGAGCGCGTAAGCTCTAGATCTCCTCGTGTTGCAACTACA"
    "CGCGCGGGTCACTCGAAACACATAGTATGAACTTAACGACTGCTCGTACTGAACAATGCTGAGGCAGAAGAT"
    "CGCAGACCAGGCATCCCACTGCTTGAAAAAACTATNNNNCTACCCGCCTTTTTATTATCTCATCAGATCAAG";

String&lt;unsigned&gt; kmerCounts;
unsigned k = 2;  // Count all 2-mers
countKmers(kmerCounts, sequenceDna5, k);

for(unsigned i = 0; i&lt;16; ++i)       // Print the 2-mer counts
    std::cout&lt;&lt;kmerCounts[i]&lt;&lt;"\n";  // 34 times AA; 30 times AC; 28 times AG; ...


String&lt;double&gt; nucleotideFrequencies;  // Defines a Bernoulli model for DNA sequences.
// Count all 2-mers and save the nucleotide frequencies
countKmers(kmerCounts, nucleotideFrequencies, sequenceDna5, k);

for(unsigned i = 0; i&lt;4; ++i)          // Print the nucleotide frequencies
    std::cout &lt;&lt; nucleotideFrequencies[i] &lt;&lt; "\n";
// =&gt; p(A) = 0.238; p(C) = 0.254; p(G) = 0.238; p(T) = 0.27;

MarkovModel&lt;Dna, double&gt;  backgroundModel(1);  // Markov model of order 1
// Count all 2-mers and return a Markov model
countKmers(kmerCounts, backgroundModel, sequenceDna5, k);
std::cout&lt;&lt;backgroundModel.transition;  // Print the transition matrix of the markov model</pre></div></div>

  </div>
</div>


<!-- @throw -->




<!-- @datarace -->
<h2 class="clause_header">Data Races</h2>

	<div class="summary compact">
 		Thread safety unknown!
	</div>


<!-- @see -->

  <h2 class="clause_header">See Also</h2>
  <ul class="summary compact">
  
  <li><a href="class_MarkovModel.html" data-lang-entity="class">MarkovModel</a></li>
  
  <li><a href="global_function_alignmentFreeComparison.html" data-lang-entity="global_function">alignmentFreeComparison</a></li>
  
  <li><a href="global_function_calculateCovariance.html" data-lang-entity="global_function">calculateCovariance</a></li>
  
  <li><a href="global_function_calculateProbability.html" data-lang-entity="global_function">calculateProbability</a></li>
  
  <li><a href="global_function_calculateVariance.html" data-lang-entity="global_function">calculateVariance</a></li>
  
  <li><a href="global_function_stringToStringSet.html" data-lang-entity="global_function">stringToStringSet</a></li>
  
  </ul>


<!--
 =============================================================================
  Footer / Debug
 =============================================================================
-->



<div class="modal fade" id="doxSources" tabindex="-1" role="dialog" aria-labelledby="doxSourcesLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
        <h4 class="modal-title" id="doxSourcesLabel">Dox Sources</h4>
      </div>
      <div class="modal-body">
        <pre>/*!
 * @fn countKmers
 * 
 * @headerfile &lt;seqan/alignment_free.h&gt;
 * 
 * @brief Counts kmers in a sequence. Optionally, a background model is
 *        returned.
 * 
 * @signature void countKmers(kmerCounts, sequence, k);
 * @signature void countKmers(kmerCounts, bgFrequencies, sequence, k);
 * @signature void countKmers(kmerCounts, bgModel, sequence, k);
 * 
 * @param[out] kmerCounts @link String @endlink of &lt;tt&gt;unsigned&lt;/tt&gt; with kmer
 *                        counts for every k-mer.
 * @param[out] bgFrequencies @link String @endlink of background frequencies
 *                           (&lt;tt&gt;double&lt;/tt&gt;) representing the model.
 * @param[out] bgModel @link MarkovModel @endlink to use.
 * @param[in] sequence @link String @endlink (sequence) where k-mers are
 *                     counted.
 * @param[in] k k-mer length (&lt;tt&gt;unsigned&lt;/tt&gt;).
 * 
 * k-mers overlapping masked (aka &#39;N&#39;) letters are not counted in case of
 * Dna5Strings. A Bernoulli or Markov Model can be choosen as a background
 * model.
 * 
 * @section Examples
 * 
 * Calculate the alignment free sequence similarity o two masked DNA sequences.
 * 
 * @code{.cpp}
 * using namespace seqan;
 * // Masked sequence, we do not want to count words overlapping &#39;N&#39;
 * Dna5String sequenceDna5 =
 *     &#34;TAGGTTTTCCGAAAAGGTAGCAACTTTACGTGATCAAACCTCTGACGGGGTTTTCCCCGTCGAAATTGGGTG&#34;
 *     &#34;TTTCTTGTCTTGTTCTCACTTGGGGCATCTCCGTCAAGCCAAGAAAGTGCTCCCTGGATTCTGTTGCTAACG&#34;
 *     &#34;AGTCTCCTCTGCATTCCTGCTTGACTGATTGGGCGGACGGGGTGTCCACCTGACGCTGAGTATCGCCGTCAC&#34;
 *     &#34;GGTGCCACATGTCTTATCTATTCAGGGATCAGAATTTATTCAGGAAATCAGGAGATGCTACACTTGGGTTAT&#34;
 *     &#34;CGAAGCTCCTTCCAAGGCGTAGCAAGGGCGACTGAGCGCGTAAGCTCTAGATCTCCTCGTGTTGCAACTACA&#34;
 *     &#34;CGCGCGGGTCACTCGAAACACATAGTATGAACTTAACGACTGCTCGTACTGAACAATGCTGAGGCAGAAGAT&#34;
 *     &#34;CGCAGACCAGGCATCCCACTGCTTGAAAAAACTATNNNNCTACCCGCCTTTTTATTATCTCATCAGATCAAG&#34;;
 * 
 * String&lt;unsigned&gt; kmerCounts;
 * unsigned k = 2;  // Count all 2-mers
 * countKmers(kmerCounts, sequenceDna5, k);
 * 
 * for(unsigned i = 0; i&lt;16; ++i)       // Print the 2-mer counts
 *     std::cout&lt;&lt;kmerCounts[i]&lt;&lt;&#34;\n&#34;;  // 34 times AA; 30 times AC; 28 times AG; ...
 * 
 * 
 * String&lt;double&gt; nucleotideFrequencies;  // Defines a Bernoulli model for DNA sequences.
 * // Count all 2-mers and save the nucleotide frequencies
 * countKmers(kmerCounts, nucleotideFrequencies, sequenceDna5, k);
 * 
 * for(unsigned i = 0; i&lt;4; ++i)          // Print the nucleotide frequencies
 *     std::cout &lt;&lt; nucleotideFrequencies[i] &lt;&lt; &#34;\n&#34;;
 * // =&gt; p(A) = 0.238; p(C) = 0.254; p(G) = 0.238; p(T) = 0.27;
 * 
 * MarkovModel&lt;Dna, double&gt;  backgroundModel(1);  // Markov model of order 1
 * // Count all 2-mers and return a Markov model
 * countKmers(kmerCounts, backgroundModel, sequenceDna5, k);
 * std::cout&lt;&lt;backgroundModel.transition;  // Print the transition matrix of the markov model
 * @endcode
 * 
 * 
 * @see alignmentFreeComparison
 * @see calculateProbability
 * @see calculateVariance
 * @see calculateCovariance
 * @see stringToStringSet
 * @see MarkovModel
 */</pre>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-primary" data-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>





    </div>

  </body>
</html>