<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" class="global_function_consensusalignment" data-page="consensusAlignment">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=US-ASCII" />
    <title>consensusAlignment - SeqAn API Documentation</title>

      

      
      <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

      <link rel="stylesheet" href="lib/bootstrap/css/bootstrap.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap/js/bootstrap.min.js"></script>

      <link rel="stylesheet" href="lib/bootstrap-multiselect/css/bootstrap-multiselect.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap-multiselect/js/bootstrap-multiselect.js"></script>

      <script type="text/javascript" charset="utf-8" src="lib/ZeroClipboard/ZeroClipboard.min.js"></script>

      <link rel="stylesheet" href="css/style.css" type="text/css" media="screen" charset="utf-8" />
      <link rel="stylesheet" href="css/common.less.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="js/less.min.js"></script>
      



    <script type="text/javascript" charset="utf-8">
      hasFrames = window.top.frames.main ? true : false;
      relpath = '/';
      docsPrefix = 'docs/yard';
      listPrefix = 'list/docs/yard';
      searchPrefix = 'search/docs/yard';
      framesUrl = '/docs/yard/frames/file/README.md';
    </script>

    <style type="text/css">
    <!-- pygments not available -->
    </style>

    <style>
      .link.error
      {
          color: red;
      }
    </style>

      
      
      <script type="text/javascript" charset="utf-8" src="lib/jquery-bbq/jquery.ba-bbq.min.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/jquery.smooth-scroll.js"></script>

      <script type="text/javascript" charset="utf-8" src="js/lang_entities.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/app.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/autocomplete.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/rubydoc_custom.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/common.js"></script>
      
  </head>
  <body>
    <a id="top" name="top"></a>

    <div id="content">
      
<h1 data-lang-entity="function" data-pimped="true">
  <a href="page_LanguageEntities.html#function">fn()</a> <span>consensusAlignment<div><div>Compute consensus alignment.</div></div></span>
</h1>


<table class="overview">

  

  

  

  

  
  </tr>

  <tr>
    <th>Defined in</th>
    <td>
      
      &lt;seqan/consensus.h&gt;
      
      
    </td>
  </tr>
  
  <tr>
    <th>Signature</th>
    <td>
      <code>void consensusAlignment(store, options);
</code>
    </td>
  </tr>
</table>


<!-- @param -->



<h2>Parameters</h2>

<table class="overview">

    <tr>
        <th>
        
            
                <code data-param-type="in_out" title="IN/OUT parameters are read and also modified by the called function">store</code>
            
        
        </th>
        <td><a href="class_FragmentStore.html" data-lang-entity="class">FragmentStore</a> to use for consensus alignment computation.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">options</code>
            
        
        </th>
        <td><a href="class_ConsensusAlignmentOptions.html" data-lang-entity="class">ConsensusAlignmentOptions</a> with configuration.</td>
    </tr>

</table>




<!-- @tparam -->





<!-- @returns -->




<!--
 =============================================================================
  Detailed Description
 =============================================================================
-->


<h2>Detailed Description</h2>
<div class="docstring">
  <div class="discussion">
    <!-- @internal -->
    

    <!-- @deprecated -->
    

    <!-- @warning -->
    

    <!-- @note -->
    

    <!-- @aka -->
    

    <!-- paragraphs -->
    
<div><p>This function computes a consensus alignment for a set of nucleic sequences that are stored in a <a href="class_FragmentStore.html" data-lang-entity="class">FragmentStore</a>. Often, consensus sequences are reads, but they could also be other sequences, such as RNA transcripts.
However, in the following description we call them reads.</p><p>This function uses the <a href="class_FragmentStore.html#FragmentStore::contigStore" data-lang-entity="member_variable">contigStore</a>, <a href="class_FragmentStore.html#FragmentStore::alignedReadStore" data-lang-entity="member_variable">alignedReadStore</a>,
and <a href="class_FragmentStore.html#FragmentStore::readSeqStore" data-lang-entity="member_variable">readSeqStore</a> members of <tt>store</tt>.</p><p>Each read must have at most one entry in <tt>store.alignedReadStore</tt>.</p><h3>General Algorithm</h3><p>In the most common case, both contig ID information and position information is available. The algorithm considers
all aligned reads on each contig. For each read, all overlapping reads (with begin/end position extended
<tt>options.posDelta</tt> to the left/to the right) are considered and overlap alignments are computed. This
pairwise alignment information is then used to compute a multiple sequence alignment (MSA).</p><p>The resulting MSA is then refined by <a href="global_function_reAlignment.html" data-lang-entity="global_function">reAlignment</a> (see <a href="class_ConsensusAlignmentOptions.html#ConsensusAlignmentOptions::runRealignment" data-lang-entity="member_variable">runRealignment</a>).</p><h3>Using position information</h3><p>When position information is to be used then this will be used to generate fewer overlap alignments by only
considering possible overlaps in windows around each read alignment as described above. Note that there can only be
at most one alignment for each read in the <tt>store.alignedReadStore</tt> and the end position must be greater than
or equal to the begin position, i.e., the alignment must be on the forward strand.</p><p>Using position information also requires contig ID information.</p><h3>Using contigID information</h3><p>When contig ID information is to be used and position information is not to be used then the
<tt>consensusAlignment()</tt> will compute pairwise alignments between all pairs of reads on the same contig.</p><p>When contigID information is to be used then only the reads with an entry in <tt>store.alignedReadStore</tt> are
considered.</p><p>When contigID information is not used then an all-to-all pairwise alignment of all reads will be performed.</p><p><strong>Note</strong> that if reads having the same contig ID cannot be properly aligned and the MSA falls apart then
the reads for one "connected alignment component" are kept on the original contig while the rest are added to new
contigs that are appended to <tt>store.contigStore</tt>.</p><h3>Example</h3><p>The following example program takes a reference sequence and creates overlapping reads from it. These are then added
to <tt>store</tt> with approximate positions (adding and subtracting a few positions). The function
<tt>consensusAlignment()</tt> is then used to compute a MSA and the consensus sequence is stored in
<tt>store.contigStore[0].seq</tt>.</p><div data-src-path="demos/dox/consensus/consensus_alignment.cpp"><pre class="code">#include &lt;iostream&gt;

#include &lt;seqan/consensus.h&gt;
#include &lt;seqan/sequence.h&gt;
#include &lt;seqan/store.h&gt;

using namespace seqan;

int main()
{
    // Reference to simulate reads from.
    Dna5String ref =
        "AATGGATGGCAAAATAGTTGTTCCATGAATACATCTCTAAAGAGCTTTGATGCTAATTTAGTCAAATTTT"
        "CAATACTGTACAATCTTCTCTAGAGCAGAGCAAAAGAATAAAAGCACTTCTAGCTAATATTATGTGGCAT";

    // Read length and step width for reads.
    int const READ_LENGTH = 50;
    int const STEP = 5;

    // Compute reads and append to FragmentStore.
    FragmentStore&lt;&gt; store;
    for (unsigned pos = 0, i = 0; pos + READ_LENGTH &lt; length(ref); pos += STEP, ++i)
    {
        // Append a new read sequence.
        unsigned readID = appendRead(store, infix(ref, pos, pos + READ_LENGTH));
        // Create small perturbation of the position but not left of position 0.
        int pos2 = std::max(0, (int)pos + ((int)i % 6 - 3));
        // Append a new read alignment for the just added read.
        appendAlignedRead(store, readID, 0, pos2, pos2 + READ_LENGTH);
    }

    // Add contig and contig name for printing.
    resize(store.contigStore, 1);
    store.contigStore[0].seq = ref;
    resize(store.contigNameStore, 1);

    // Print initial perturbated alignment.
    std::cout &lt;&lt; "Initial Alignment\n\n";
    AlignedReadLayout layout;
    layoutAlignment(layout, store);
    printAlignment(std::cout, layout, store, /*contigID=*/ 0,
                   /*beginPos=*/ 0, /*endPos=*/ (int)length(ref), 0, 30);

    // Compute consensus alignment.
    ConsensusAlignmentOptions options;
    consensusAlignment(store, options);

    // Print final consensus alignment.
    std::cout &lt;&lt; "\n\nFinal Consensus Alignment\n\n";
    layoutAlignment(layout, store);
    printAlignment(std::cout, layout, store, /*contigID=*/ 0,
                   /*beginPos=*/ 0, /*endPos=*/ (int)length(ref), 0, 30);

    return 0;
}
</pre><div class="path_label"><span class="label">Demo:</span> <a href="demos/dox/consensus/consensus_alignment.cpp" target="_top">demos/dox/consensus/consensus_alignment.cpp</a></div></div><p>The output is as follows:</p><pre class="console" data-src-path="demos/dox/consensus/consensus_alignment.cpp.stdout">Initial Alignment

AATGGATGGCAAAATAGTTGTTCCATGAATACATCTCTAAAGAGCTTTGATGCTAATTTAGTCAAATTTTCAATACTGTACAATCTTCTCTAGAGCAGAGCAAAAGAATAAAAGCACTTCTAGCTAATATTATGTGGCAT
.................................................. TGCT.A..TAGTC..A...TC.ATACTGTAC.ATC.TCTCTAGAGCAGAG
   ATGGCAA..T.GTTG.TCCATGAAT.C.TC...AA.G..CTT.GATGCTA    A.TTAGTC.AA..TTCA.TAC.GT.CAA.CT...CT..AGC..AGC..A.
         A...TAGT.GT.C.ATGA.TACATCTCTA..GAGCT..GATGCTA.T..A    .TT..CAAT.CTG.AC.ATC.TC...AG..CAG..CAA..G.AT.A..GC
               ..................................................    CA.TACTGTACA.TCT.CTCTAGAGCAGAGCA...GA.TA...GCACT.C
                     .T.CATG.ATACATCTCT..AGAGC..TGATGCT.A..TAGTC..A...T    ..................................................
                           TGA.T.CA...CT.A..AGC.TTGATGCTAA.TTAGTC.AA..TTCA.TA    C.ATC.TCTCTAGAGCAGAGC...AG.AT...AGCAC.TCTAGCT.ATA.
                           .CAT.TCTAAAG.GCTTTGA..C.AA.TTAG.CAAATTT.CAAT.C.GTA          T...CT..AGC..AGC..A.GA.T..AAG.ACT.CTAGCTA..AT.A..T
                                 ...AA.G..CTT.GATGCTAATT.AGTCAA.TT..CAAT.CTG.AC.ATC
                                       .GAGCT..GATGCTA.T..AGTCA..T...CA.TACTGTACA.TCT.CTC
                                             ..................................................
                                                         G.CAAATTT.CAAT.C.GTACA.TCT...CTAGAGC..AGC.AA.G..T.


Final Consensus Alignment

AATGGATGGCAAAATAGTTGTTCCATGAATACATCTCTAAAGAGCTTTGATGCTAATTTAGTCAAATTTTCAATACTGTACAATCTTCTCTAGAGCAGAGCAAAAGAATAAAAGCACTTCTAGCTAATATTATGT-----
..................................................     ..................................................
     ..................................................     ..................................................
          ..................................................     ..................................................
               ..................................................     ..................................................
                    ..................................................     ..................................................
                         ..................................................     ..................................................
                              ..................................................     ..................................................
                                   ..................................................
                                        ..................................................
                                             ..................................................
                                                  ..................................................
</pre></div>

  </div>
</div>


<!-- @throw -->


<h2>Thrown Exceptions</h2>

<table class="overview">

    <tr>
        <th><code>s</code></th>
        <td>ConsensusAlignerIllegalArgumentException in case of invalid arguments (e.g. two alignments for the same
read).</td>
    </tr>

</table>



<!-- @datarace -->
<h2 class="clause_header">Data Races</h2>

	<div class="summary compact">
 		Thread safety unknown!
	</div>


<!-- @see -->


<!--
 =============================================================================
  Footer / Debug
 =============================================================================
-->



<div class="modal fade" id="doxSources" tabindex="-1" role="dialog" aria-labelledby="doxSourcesLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
        <h4 class="modal-title" id="doxSourcesLabel">Dox Sources</h4>
      </div>
      <div class="modal-body">
        <pre>/*!
 * @fn consensusAlignment
 * 
 * @headerfile &lt;seqan/consensus.h&gt;
 * 
 * @brief Compute consensus alignment.
 * 
 * @signature void consensusAlignment(store, options);
 * 
 * @param[in,out] store @link FragmentStore @endlink to use for consensus
 *                      alignment computation.
 * @param[in] options @link ConsensusAlignmentOptions @endlink with
 *                    configuration.
 * 
 * @throw s ConsensusAlignerIllegalArgumentException in case of invalid
 *          arguments (e.g. two alignments for the same read).
 * 
 * This function computes a consensus alignment for a set of nucleic sequences
 * that are stored in a @link FragmentStore @endlink. Often, consensus sequences
 * are reads, but they could also be other sequences, such as RNA transcripts.
 * However, in the following description we call them reads.
 * 
 * This function uses the @link FragmentStore::contigStore @endlink, @link
 * FragmentStore::alignedReadStore @endlink, and @link
 * FragmentStore::readSeqStore @endlink members of &lt;tt&gt;store&lt;/tt&gt;.
 * 
 * Each read must have at most one entry in &lt;tt&gt;store.alignedReadStore&lt;/tt&gt;.
 * 
 * @section General Algorithm
 * 
 * In the most common case, both contig ID information and position information
 * is available. The algorithm considers all aligned reads on each contig. For
 * each read, all overlapping reads (with begin/end position extended
 * &lt;tt&gt;options.posDelta&lt;/tt&gt; to the left/to the right) are considered and
 * overlap alignments are computed. This pairwise alignment information is then
 * used to compute a multiple sequence alignment (MSA).
 * 
 * The resulting MSA is then refined by @link reAlignment @endlink (see @link
 * ConsensusAlignmentOptions::runRealignment @endlink).
 * 
 * @section Using position information
 * 
 * When position information is to be used then this will be used to generate
 * fewer overlap alignments by only considering possible overlaps in windows
 * around each read alignment as described above. Note that there can only be at
 * most one alignment for each read in the &lt;tt&gt;store.alignedReadStore&lt;/tt&gt; and
 * the end position must be greater than or equal to the begin position, i.e.,
 * the alignment must be on the forward strand.
 * 
 * Using position information also requires contig ID information.
 * 
 * @section Using contigID information
 * 
 * When contig ID information is to be used and position information is not to
 * be used then the &lt;tt&gt;consensusAlignment()&lt;/tt&gt; will compute pairwise
 * alignments between all pairs of reads on the same contig.
 * 
 * When contigID information is to be used then only the reads with an entry in
 * &lt;tt&gt;store.alignedReadStore&lt;/tt&gt; are considered.
 * 
 * When contigID information is not used then an all-to-all pairwise alignment
 * of all reads will be performed.
 * 
 * &lt;strong&gt;Note&lt;/strong&gt; that if reads having the same contig ID cannot be
 * properly aligned and the MSA falls apart then the reads for one &#34;connected
 * alignment component&#34; are kept on the original contig while the rest are added
 * to new contigs that are appended to &lt;tt&gt;store.contigStore&lt;/tt&gt;.
 * 
 * @section Example
 * 
 * The following example program takes a reference sequence and creates
 * overlapping reads from it. These are then added to &lt;tt&gt;store&lt;/tt&gt; with
 * approximate positions (adding and subtracting a few positions). The function
 * &lt;tt&gt;consensusAlignment()&lt;/tt&gt; is then used to compute a MSA and the consensus
 * sequence is stored in &lt;tt&gt;store.contigStore[0].seq&lt;/tt&gt;.
 * 
 * @include demos/dox/consensus/consensus_alignment.cpp
 * 
 * The output is as follows:
 * 
 * @include demos/dox/consensus/consensus_alignment.cpp.stdout
 */</pre>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-primary" data-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>





    </div>

  </body>
</html>