<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en" class="global_function_realignment" data-page="reAlignment">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=US-ASCII" />
    <title>reAlignment - SeqAn API Documentation</title>

      

      
      <script type="text/javascript" charset="utf-8" src="js/jquery.js"></script>

      <link rel="stylesheet" href="lib/bootstrap/css/bootstrap.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap/js/bootstrap.min.js"></script>

      <link rel="stylesheet" href="lib/bootstrap-multiselect/css/bootstrap-multiselect.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="lib/bootstrap-multiselect/js/bootstrap-multiselect.js"></script>

      <script type="text/javascript" charset="utf-8" src="lib/ZeroClipboard/ZeroClipboard.min.js"></script>

      <link rel="stylesheet" href="css/style.css" type="text/css" media="screen" charset="utf-8" />
      <link rel="stylesheet" href="css/common.less.css" type="text/css" media="screen" charset="utf-8" />
      <script type="text/javascript" charset="utf-8" src="js/less.min.js"></script>
      



    <script type="text/javascript" charset="utf-8">
      hasFrames = window.top.frames.main ? true : false;
      relpath = '/';
      docsPrefix = 'docs/yard';
      listPrefix = 'list/docs/yard';
      searchPrefix = 'search/docs/yard';
      framesUrl = '/docs/yard/frames/file/README.md';
    </script>

    <style type="text/css">
    <!-- pygments not available -->
    </style>

    <style>
      .link.error
      {
          color: red;
      }
    </style>

      
      
      <script type="text/javascript" charset="utf-8" src="lib/jquery-bbq/jquery.ba-bbq.min.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/jquery.smooth-scroll.js"></script>

      <script type="text/javascript" charset="utf-8" src="js/lang_entities.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/app.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/autocomplete.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/rubydoc_custom.js"></script>
      <script type="text/javascript" charset="utf-8" src="js/common.js"></script>
      
  </head>
  <body>
    <a id="top" name="top"></a>

    <div id="content">
      
<h1 data-lang-entity="function" data-pimped="true">
  <a href="page_LanguageEntities.html#function">fn()</a> <span>reAlignment<div><div>Perform realignment on a <a href="class_FragmentStore.html" data-lang-entity="class">FragmentStore</a> object.</div></div></span>
</h1>


<table class="overview">

  

  

  

  

  
  </tr>

  <tr>
    <th>Defined in</th>
    <td>
      
      &lt;seqan/realign.h&gt;
      
      
    </td>
  </tr>
  
  <tr>
    <th>Signature</th>
    <td>
      <code>void reAlignment(store, contigID, realignmentMethod, bandwidth, includeReference[,
windowBegin, windowEnd][, debug][, printTiming]);
</code>
    </td>
  </tr>
</table>


<!-- @param -->



<h2>Parameters</h2>

<table class="overview">

    <tr>
        <th>
        
            
                <code data-param-type="in_out" title="IN/OUT parameters are read and also modified by the called function">store</code>
            
        
        </th>
        <td>The <a href="class_FragmentStore.html" data-lang-entity="class">FragmentStore</a> to perform realignment for.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">realignmentMethod</code>
            
        
        </th>
        <td>The realignment algorithm to use, <tt>1</tt> for affine gap costs, <tt>0</tt>
for linear gap costs, affine gap costs are recommended, <tt>unsigned</tt>.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">contigID</code>
            
        
        </th>
        <td>Identifier of the contig to realign, <tt>unsigned</tt></td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">bandwidth</code>
            
        
        </th>
        <td>Bandwidth to use in the pairwise DP alignment algorithms <tt>unsigned</tt>.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">includeReference</code>
            
        
        </th>
        <td>A <tt>bool</tt>, if <tt>true</tt> then the reference will be included as a
pseudo-read to increase stability in case of indels. See section "Including
the Reference" for details.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">windowBegin</code>
            
        
        </th>
        <td>Optional <tt>unsigned</tt> window begin for windowed alignment,
default: <tt>0</tt>, i.e. no windowing. Also see section "Windowed
Realignment" below.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">windowEnd</code>
            
        
        </th>
        <td>Optional <tt>unsigned</tt> window end for windowed alignment, default: <tt>0</tt>,
i.e. no windowing. Also see section "Windowed Realignment" below.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">debug</code>
            
        
        </th>
        <td>Optional <tt>bool</tt> to enable verbose logging, default: <tt>false</tt>.</td>
    </tr>

    <tr>
        <th>
        
            
                <code data-param-type="in" title="IN parameters are only read and not modified">printTiming</code>
            
        
        </th>
        <td>Optional <tt>bool</tt> to enable printing of times, default: <tt>false</tt>.</td>
    </tr>

</table>




<!-- @tparam -->





<!-- @returns -->




<!--
 =============================================================================
  Detailed Description
 =============================================================================
-->


<h2>Detailed Description</h2>
<div class="docstring">
  <div class="discussion">
    <!-- @internal -->
    

    <!-- @deprecated -->
    

    <!-- @warning -->
    

    <!-- @note -->
    

    <!-- @aka -->
    

    <!-- paragraphs -->
    
<div><p>This function implements a variant of the Anson-Myers algorithm to refine the multi-read alignments. The algorithm
works in a round-robin fashion:</p><p>First, the profile sequence is created for the multi-read alignment. The algorithm then works in round.</p><p>In each round, each read alignment is selected, removed from the profile and then aligned against the profile using
a DP algorithm. For this alignment, a region of <tt>bandwidth / 2</tt> positions around the previous alignment
locus is chosen and aligned using a bandwidth of <tt>bandwidth</tt> in the DP alignment algorithm. The profile is
then updated with the resulting pairwise alignment. This is iterated until the global alignment score does not
improve.</p><p>In the end, the reference sequence of the given contig is replaced by the consensus of the multi-read alignment.</p><h3>Including the Reference</h3><p>You can choose to include the reference sequence of the selected contig as a pseudo-read. In this case, the
reference sequence of the selected contig will be added as an additional read and also a read alignment will be
created for it. The appended pseudo-read will have the highest read ID in the store.</p><h3>Windowed Realignment</h3><p>Sometimes, it is useful to just realign a part of the fragment store. This can be done by using the parameters
<tt>windowBegin</tt> and <tt>windowEnd</tt>. This refers to view positions in the current multi-read alignment.
Set both to <tt>0</tt> (also the default) to disalbe windowed realignment.</p><h3>References</h3><p><ul><li>Anson, Eric L., and Eugene W. Myers. "ReAligner: a program for refining DNA sequence multi-alignments."
Journal of Computational Biology 4.3 (1997): 369-383.</li></ul></p></div>

  </div>
</div>


<!-- @throw -->




<!-- @datarace -->
<h2 class="clause_header">Data Races</h2>

	<div class="summary compact">
 		Thread safety unknown!
	</div>


<!-- @see -->


<!--
 =============================================================================
  Footer / Debug
 =============================================================================
-->



<div class="modal fade" id="doxSources" tabindex="-1" role="dialog" aria-labelledby="doxSourcesLabel" aria-hidden="true">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close" data-dismiss="modal" aria-hidden="true">&times;</button>
        <h4 class="modal-title" id="doxSourcesLabel">Dox Sources</h4>
      </div>
      <div class="modal-body">
        <pre>/*!
 * @fn reAlignment
 * 
 * @headerfile &lt;seqan/realign.h&gt;
 * 
 * @brief Perform realignment on a @link FragmentStore @endlink object.
 * 
 * @signature void reAlignment(store, contigID, realignmentMethod, bandwidth,
 *            includeReference[, windowBegin, windowEnd][, debug][,
 *            printTiming]);
 * 
 * @param[in,out] store The @link FragmentStore @endlink to perform realignment
 *                      for.
 * @param[in] realignmentMethod The realignment algorithm to use, &lt;tt&gt;1&lt;/tt&gt; for
 *                              affine gap costs, &lt;tt&gt;0&lt;/tt&gt; for linear gap
 *                              costs, affine gap costs are recommended,
 *                              &lt;tt&gt;unsigned&lt;/tt&gt;.
 * @param[in] contigID Identifier of the contig to realign, &lt;tt&gt;unsigned&lt;/tt&gt;
 * @param[in] bandwidth Bandwidth to use in the pairwise DP alignment algorithms
 *                      &lt;tt&gt;unsigned&lt;/tt&gt;.
 * @param[in] includeReference A &lt;tt&gt;bool&lt;/tt&gt;, if &lt;tt&gt;true&lt;/tt&gt; then the
 *                             reference will be included as a pseudo-read to
 *                             increase stability in case of indels. See section
 *                             &#34;Including the Reference&#34; for details.
 * @param[in] windowBegin Optional &lt;tt&gt;unsigned&lt;/tt&gt; window begin for windowed
 *                        alignment, default: &lt;tt&gt;0&lt;/tt&gt;, i.e. no windowing.
 *                        Also see section &#34;Windowed Realignment&#34; below.
 * @param[in] windowEnd Optional &lt;tt&gt;unsigned&lt;/tt&gt; window end for windowed
 *                      alignment, default: &lt;tt&gt;0&lt;/tt&gt;, i.e. no windowing. Also
 *                      see section &#34;Windowed Realignment&#34; below.
 * @param[in] debug Optional &lt;tt&gt;bool&lt;/tt&gt; to enable verbose logging, default:
 *                  &lt;tt&gt;false&lt;/tt&gt;.
 * @param[in] printTiming Optional &lt;tt&gt;bool&lt;/tt&gt; to enable printing of times,
 *                        default: &lt;tt&gt;false&lt;/tt&gt;.
 * 
 * This function implements a variant of the Anson-Myers algorithm to refine the
 * multi-read alignments. The algorithm works in a round-robin fashion:
 * 
 * First, the profile sequence is created for the multi-read alignment. The
 * algorithm then works in round.
 * 
 * In each round, each read alignment is selected, removed from the profile and
 * then aligned against the profile using a DP algorithm. For this alignment, a
 * region of &lt;tt&gt;bandwidth / 2&lt;/tt&gt; positions around the previous alignment
 * locus is chosen and aligned using a bandwidth of &lt;tt&gt;bandwidth&lt;/tt&gt; in the DP
 * alignment algorithm. The profile is then updated with the resulting pairwise
 * alignment. This is iterated until the global alignment score does not
 * improve.
 * 
 * In the end, the reference sequence of the given contig is replaced by the
 * consensus of the multi-read alignment.
 * 
 * @section Including the Reference
 * 
 * You can choose to include the reference sequence of the selected contig as a
 * pseudo-read. In this case, the reference sequence of the selected contig will
 * be added as an additional read and also a read alignment will be created for
 * it. The appended pseudo-read will have the highest read ID in the store.
 * 
 * @section Windowed Realignment
 * 
 * Sometimes, it is useful to just realign a part of the fragment store. This
 * can be done by using the parameters &lt;tt&gt;windowBegin&lt;/tt&gt; and
 * &lt;tt&gt;windowEnd&lt;/tt&gt;. This refers to view positions in the current multi-read
 * alignment. Set both to &lt;tt&gt;0&lt;/tt&gt; (also the default) to disalbe windowed
 * realignment.
 * 
 * @section References
 * 
 * &lt;ul&gt;&lt;li&gt;Anson, Eric L., and Eugene W. Myers. &#34;ReAligner: a program for
 * refining DNA sequence multi-alignments.&#34; Journal of Computational Biology 4.3
 * (1997): 369-383.&lt;/li&gt;&lt;/ul&gt;
 */</pre>
      </div>
      <div class="modal-footer">
        <button type="button" class="btn btn-primary" data-dismiss="modal">Close</button>
      </div>
    </div>
  </div>
</div>





    </div>

  </body>
</html>